server/core/region.go

```diff
@@ -651,16 +689,65 @@ func (r *RegionsInfo) AddRegion(region *RegionInfo) []*RegionInfo {
 		}
 		store.update(region)
 	}
+}
 
-	return overlaps
+// shouldRemoveFromSubTree return true when the region leader changed, peer transferred,
+// new peer was created, learners changed, pendingPeers changed, and so on.
+func (r *RegionsInfo) shouldRemoveFromSubTree(region *RegionInfo, origin *RegionInfo) bool {
+	checkPeersChange := func(origin []*metapb.Peer, other []*metapb.Peer) bool {
+		for _, a := range origin {
```

Comment:
The code between line 698 and line 716 has some overlap logic?

Reply:
Do some change, ok?
```
	checkPeersChange := func(origin []*metapb.Peer, other []*metapb.Peer) bool {
		checkPeersDiff := func(origin []*metapb.Peer, other []*metapb.Peer) bool {
			for _, a := range origin {
				if a == nil {
					continue
				}
				both := false
				for _, b := range other {
					if b == nil {
						continue
					}
					if a.GetStoreId() == b.GetStoreId() && a.GetId() == b.GetId() {
						both = true
						break
					}
				}
				if !both {
					return true
				}
			}
			return false
		}
		if checkPeersDiff(origin, other) || checkPeersDiff(other, origin) {
			return true
		}
		return false
	}
```